File: backend/app/modules/diets/nutrition_goal.py
"""Derive nutrition goal from current vs target body weight."""

from typing import Literal

DietGoal = Literal["loss", "gain", "maintain"]

WEIGHT_TOLERANCE_KG = 0.5

GOAL_LABELS: dict[DietGoal, str] = {
    "loss": "Weight loss (cutting)",
    "gain": "Weight gain",
    "maintain": "Maintenance",
}


def derive_nutrition_goal(
    current_kg: float | None,
    goal_kg: float | None,
) -> DietGoal:
    if current_kg is None or goal_kg is None:
        return "maintain"
    if abs(current_kg - goal_kg) <= WEIGHT_TOLERANCE_KG:
        return "maintain"
    if current_kg < goal_kg:
        return "gain"
    return "loss"


def nutrition_goal_label(goal: DietGoal) -> str:
    return GOAL_LABELS.get(goal, "Maintenance")

File: frontend/src/lib/nutrition-goal.ts
export type NutritionGoal = "loss" | "gain" | "maintain";

const WEIGHT_TOLERANCE_KG = 0.5;

export const NUTRITION_GOAL_LABELS: Record<NutritionGoal, string> = {
  loss: "Weight loss (cutting)",
  gain: "Weight gain",
  maintain: "Maintenance",
};

export function deriveNutritionGoal(
  currentKg?: number | null,
  goalKg?: number | null,
): NutritionGoal {
  if (currentKg == null || goalKg == null) {
    return "maintain";
  }
  if (Math.abs(currentKg - goalKg) <= WEIGHT_TOLERANCE_KG) {
    return "maintain";
  }
  if (currentKg < goalKg) {
    return "gain";
  }
  return "loss";
}

export function nutritionGoalLabel(goal: NutritionGoal): string {
  return NUTRITION_GOAL_LABELS[goal] ?? "Maintenance";
}
